ouroboros-consensus-cardano/tools/db-analyser/Main.hs

```diff
@@ -1,307 +1,97 @@
-              , show (blockPrevHash (configCodec cfg) blk)
-              , show (    Map.lookup
-                            (blockHash blk)
-                            (Analysis.knownEBBs (Proxy @blk))
-                       == Just (blockPrevHash (configCodec cfg) blk)
-                     )
-              ]
-          _otherwise ->
-            return () -- Skip regular blocks
+data CmdLine = CmdLine {
+    dbDir      :: FilePath
+  , verbose    :: Bool
+  , onlyImmDB  :: Bool
+  , validation :: Maybe ValidateBlocks
+  , blockType  :: BlockType
+  , analysis   :: AnalysisName
+  }
 
-{-------------------------------------------------------------------------------
-  Auxiliary: processing all blocks in the imm DB
--------------------------------------------------------------------------------}
+data ValidateBlocks = ValidateAll | MinimumValidation
 
-processAll :: forall blk. (HasHeader blk, ImmDbSerialiseConstraints blk)
-           => ImmDB IO blk
-           -> ResourceRegistry IO
-           -> (blk -> IO ())
-           -> IO ()
-processAll immDB rr callback = do
-    tipPoint <- getPointAtTip immDB
-    case pointToWithOriginRealPoint tipPoint of
-      Origin -> return ()
-      At tip -> do
-        Right itr <- stream immDB rr GetBlock
-          (StreamFromExclusive genesisPoint)
-          (StreamToInclusive tip)
-        go itr
-  where
-    go :: Iterator (HeaderHash blk) IO (IO blk) -> IO ()
-    go itr = do
-        itrResult <- ImmDB.iteratorNext itr
-        case itrResult of
-          IteratorExhausted   -> return ()
-          IteratorResult mblk -> mblk >>= \blk -> callback blk >> go itr
+data BlockType =
+    ByronBlock   ByronBlockArgs
+  | ShelleyBlock ShelleyBlockArgs
+  | CardanoBlock CardanoBlockArgs
 
 {-------------------------------------------------------------------------------
-  Command line args
+  Parsing
 -------------------------------------------------------------------------------}
 
-data CmdLine = CmdLine {
-      clConfig    :: [FilePath]
-    , clIsMainNet :: Bool
-    , clImmDB     :: FilePath
-    , clBlockType :: Maybe BlockType
-    , clAnalysis  :: AnalysisName
-    } deriving Show
-
-data BlockType = Byron | Shelley | Cardano
-  deriving Show
-
 parseCmdLine :: Parser CmdLine
 parseCmdLine = CmdLine
-    <$> many (strOption (mconcat [
-            long "config"
-          , help "Path to config file or files. Multiple occurences of the\
-                 \option can be used"
-          , metavar "PATH"
-          ]))
-    <*> flag True False (mconcat [
-            long "testnet"
-          , help "The DB contains blocks from testnet rather than mainnet"
-          ])
-    <*> strOption (mconcat [
+    <$> strOption (mconcat [
             long "db"
-          , help "Path to the chain DB (parent of \"immutable\" directory)"
+          , help "Path to the chain DB"
           , metavar "PATH"
           ])
-    <*> parseBlockType
+    <*> flag False True (mconcat [
```

Comment:
Use [`switch`](http://hackage.haskell.org/package/optparse-applicative-0.16.0.0/docs/Options-Applicative.html#v:switch), same for the flag below.